File: javascript/leaky-integrator.js
export default class LI {
	/**
        * Creates a new Leaky Integrator
        * @param {number} numInputs, the number of inputs the LI has.
    */
	constructor(numInputs) {
		// Get set when configuration loaded.
		this.gain; // Spread of curve. High gain = step function.
		this.bias; // Shift left right.
		this.t; // Time constant like spring. Higher = faster.
		this.sineCoefficient; // For sin transfer func (amplitude).
		this.frequencyMultiplier; // For sin transfer fund (frequency).
		this.transferFunction; // Gets set later.
		this.y = 0; // Activation to pass into transfer function with gain and bias.
		this.timeStep = 0;
		this.inputs = new Array(numInputs);
		this.weights = new Array(numInputs);
		this.numInputs = numInputs;
		this.tempOutput = 0;
		this.output = 0;
	}
	/**
        * Calculates the temp output of LI.
    */
	calculateTempOutput() {
		let y_dot;
		y_dot = -1.0 * this.y;
		// g(y-b) already calculated for inputs.
		for (let k = 0; k < this.numInputs; k++) {
			y_dot += this.weights[k] * this.inputs[k];
		}
		y_dot /= this.t;
		this.y += y_dot * this.timeStep;
		this.tempOutput = this.transferFunction(this.gain * (this.y - this.bias), this.sineCoefficient, this.frequencyMultiplier);
	}
	/**
        * Assigns temp output to actual output, setting neuron state.
    */
	update() {
		this.output = this.tempOutput;
	}
	/**
        * Gets the output of the LI.
    */	
   	getOutput() {
		return this.output;
	}
	/**
        * Resets the LI.
    */	
	reset() {
		this.y = 0;
		this.output = 0;
		this.tempOutput = 0;
	}
}

File: java/LeakyIntegrator.java
package com.steffanianigro.plecto;

public class LeakyIntegrator {

    // Passed into constructor.
    private int numInputs;
    public float timeStep;
    private float timeConstant;
    public float gain;
    private float bias;
    private float sineCoefficient;
    private float frequencyMultiplier;
    // Get initialised in constructor.
    public float inputs[];
    public float weights[];
    private Params params;
    // Params change over time.
    public float y = 0f;
    private float tempOutput = 0f;
    public float output = 0f;

    LeakyIntegrator(Params params, int numInputs, float timeStep, float timeConstant, float gain, float bias, float sineCoefficient, float frequencyMultiplier) {
        this.timeStep = timeStep;
        this.timeConstant = timeConstant;
        this.gain = gain;
        this.bias = bias;
        this.sineCoefficient = sineCoefficient;
        this.frequencyMultiplier = frequencyMultiplier;
        this.params = params;
        this.inputs = new float[numInputs];
        this.weights = new float[numInputs];
        this.numInputs = numInputs;

    }

    public void calculateOutput() {
        float yDot = -1.0f * this.y;
        for(int k = 0; k < this.numInputs; k++){
            yDot += this.weights[k] * this.inputs[k];
        }
        yDot /= this.timeConstant;
        this.y += yDot * this.timeStep;
        this.tempOutput = this.params.transferFunction(this.gain * (this.y - this.bias), this.sineCoefficient, this.frequencyMultiplier);
    }


    public void updateTimeStep(float timeStep){
        this.timeStep = timeStep;
    }

    public void update(){
        this.output = this.tempOutput;
    }

    public float getOutput(){
        return this.output;
    }

    public void reset() {
        this.y = 0;
        this.output = 0;
        this.tempOutput = 0;
    }

}
